clevertap-core/src/main/java/com/clevertap/android/sdk/network/api/CtApi.kt

```diff
@@ -41,74 +46,192 @@ internal class CtApi(
+            defaultHeaders.plus(HEADER_CUSTOM_HANDSHAKE to customHandshakeDomain!!)
+        } else {
+            defaultHeaders
+        }
         val request = createRequest(
-            relativePath = "hello",
+            baseUrl = baseUrl,
+            relativeUrl = "hello",
             body = null,
-            useSpikyDomain = useSpikyDomain,
-            includeTs = false
+            includeTs = false,
+            headers = headers
         )
+
         logger.verbose(logTag, "Performing handshake with ${request.url}")
+
         return httpClient.execute(request)
     }
 
     fun defineVars(body: SendQueueRequestBody): Response =
         httpClient.execute(
-            createRequest("defineVars", body.toString(), useSpikyDomain = false, includeTs = true)
+            createRequest(
+                baseUrl = getActualDomain(isViewedEvent = false) ?: defaultDomain,
+                relativeUrl = "defineVars",
+                body = body.toString()
+            )
         )
 
     fun defineTemplates(body: DefineTemplatesRequestBody): Response =
         httpClient.execute(
-            createRequest("defineTemplates", body.toString(), useSpikyDomain = false, includeTs = true)
+            createRequest(
+                baseUrl = getActualDomain(isViewedEvent = false) ?: defaultDomain,
+                relativeUrl = "defineTemplates",
+                body = body.toString()
+            )
         )
 
-    fun getActualDomain(useSpikyDomain: Boolean): String? {
+    fun getActualDomain(isViewedEvent: Boolean): String? {
         return when {
             !region.isNullOrBlank() -> {
-                val regionSuffix = if (useSpikyDomain) spikyRegionSuffix else ""
-                "$region${regionSuffix}.$defaultDomain"
+                buildString {
+                    append(region)
+                    append(
+                        if (isViewedEvent) {
+                            spikyRegionSuffix
+                        } else {
+                            ""
+                        }
+                    )
+                    append(".")
+                    append(defaultDomain)
+                }
             }
 
-            !useSpikyDomain && !proxyDomain.isNullOrBlank() -> {
+            !isViewedEvent && !proxyDomain.isNullOrBlank() -> {
                 proxyDomain
             }
 
-            useSpikyDomain && !spikyProxyDomain.isNullOrBlank() -> {
+            isViewedEvent && !spikyProxyDomain.isNullOrBlank() -> {
                 spikyProxyDomain
             }
 
-            else -> if (useSpikyDomain) {
-                spikyDomain
+            else -> if (isViewedEvent) {
+                cachedSpikyDomain
             } else {
-                domain
+                cachedDomain
+            }
+        }
+    }
+
+    fun getHandshakeDomain(isViewedEvent: Boolean) : String {
+        if (region.isNotNullAndBlank()) {
```

Comment:
Why do we need below code? `needsHandshake()` will always return false when region, spikyProxyDomain, ProxyDomain, cachedSpikyDomain and cachedDomain is `isNotNullAndBlank` in which case this will never executed. Also if we are keeping it for re-usability then these are not valid handshake domains. IMO `$defaultDomain/hello` and `customHandshakeDomain` are only valid handshake urls.
```
if (region.isNotNullAndBlank()) {
            return buildString {
                append(region)
                append(
                    if (isViewedEvent) {
                        spikyRegionSuffix
                    } else {
                        ""
                    }
                )
                append(".")
                append(defaultDomain)
            }
        }

        if (isViewedEvent) {
            if (spikyProxyDomain.isNotNullAndBlank()) {
                return spikyProxyDomain!!
            }
        } else {
            if (proxyDomain.isNotNullAndBlank()) {
                return proxyDomain!!
            }
        }

        if (isViewedEvent) {
            if (cachedSpikyDomain.isNotNullAndBlank()) {
                return cachedSpikyDomain!!
            }
        } else {
            if (cachedDomain.isNotNullAndBlank()) {
                return cachedDomain!!
            }
        }

```

Reply:
This is legacy code and there is a potential configuration(misconfiguration) from client side where they only add either of proxy/spiky-proxy domains.